# Homework 1: Python fundamentals
## And using external resources to better undersand code!

### <p style="text-align: right;"> &#9989; **Mark Endicott** </p>

## Learning Goals

### Content Goals
- Review the arithmetic operators in Python
- Utilize loops to perform repeated operations and to store, access, and manipulate data in one more or lists
- Utilize variables and lists to store and access data
- Use conditional statements to control code flow.

### Practice Goals
- Use the internet as a resource for learning how to write code, learn new things, and troubleshoot problems
- Solve a variety of problems using code
- Troubleshoot errors and debug your code as issues arise (you'll do this all the time in this course!)
___



## Assignment instructions

Work through the following assignment, making sure to follow all the directions and answer all the questions.

**This assignment is due at 11:59 pm on Friday, February 2.** It should be uploaded into the "Homework Assignments" submission folder for Homework #1.  Submission instructions can be found at the end of the notebook.


---
## Version check

Before you begin, make sure your Jupyter kernel is Python 3. There are significant differences between Python 2 and Python 3, and we use Python 3 in this course. The kernel type at the upper-right corner should say "Python 3", and the following code should show a version of the form: `3.x.y` (where "x" and "y" are the sub-version numbers).

If your kernel is Python 2, try selecting Python 3 from the Kernel menu above under the sub-menu "Change kernel". If that does not work, consult the Teams help channel for assistance.

In [1]:
import sys
print(sys.version)

3.10.9 | packaged by Anaconda, Inc. | (main, Mar  1 2023, 18:18:15) [MSC v.1916 64 bit (AMD64)]


---
<a id="toc"></a>

## Table of Contents

[Part 0. Integrity statement and office hours](#part_0) (4 point)

[Part 1. Doing basic math with Python](#part_1)  (6 points)

[Part 2. Finding common divisors](#part_2) (4 points)

[Part 3. Fibonacci numbers](#part_3)  (14 points)

[Part 4. Amortizing loans](#part_4) (16 points)

[Part 5. Dictionaries and everything else](#part_5) (7 points)

[Part 6. Conditional statements and debugging](#part_6) (6 points)

In [2]:
points_list = [4, 6, 4, 14, 16, 7, 6]

print(f"The total number of points is : {sum(points_list)}")

The total number of points is : 57


---
<a id="part_0"></a>

## Part 0: Integrity statement and office hours

[Back to Top](#toc)

### 0.1 Academic integrity statement (2 Points)

In the markdown cell below, paste your personal academic integrity statement. **You should have generated this in the Day 6 pre-class assignment**, if you did not do this pre-class assignment, locate that assignment, find the appropriate section, and draft your integrity statment now.

By including this statement, you are confirming that you are submitting this as your own work and not that of someone else.


As a Spartan, I will strive to uphold values of the highest ethical standard. I will practice honesty in my work, foster honesty in my peers, and take pride in knowing that honor in ownership is worth more than grades. I will carry these values beyond my time as a student at Michigan State University, continuing the endeavor to build personal integrity in all that I do.

### 0.2 Going to Office/Help Room Hours (2 Points)

#### Why are we doing this?
We want to make sure that everyone knows how to access the resources
available to you. One of the best resources you have at your disposal is office/help room hours, which are offered by the instructors, TAs, and LAs associated with CMSE 201.

#### What will you do?
(At minimum) Go to one office hour session ​(it doesn’t matter which one you go to). Come with one question that you would like to talk about. It can be big or small. Ask your question. All of the instructional staff for CMSE 201 (section instructors, TAs, and LAs) will have access to a roster of the folks enrolled in CMSE 201 and will be able to mark off that you stopped in. As long as there ends up being a check next to your name, you’ll get credit for this part of Homework 1.

If your schedule is such that none of the nearly 40 hours of office hours/help room work with your schedule. Send and email to your instructor to coordinate an option to make sure you can get credit for this part of the assignment.

**NOTE:** The day when the homework is due (**Friday, February 2nd**) will be the busiest time for folks to go to office hours. You are **STRONGLY** encouraged to go to office hours before that Friday to get credit for this part of this assignment. If you show up to office/help room hours on that Friday and it's very busy, there is no guarantee that one of the instructors, TAs, or LAs will be able to check you off for this part of the assignment. (You should still feel free to go to office hours on Friday for help, though!)

You can find the office hours calendar on the [course website](https://cmse.msu.edu/cmse201). There is a link to the calendar on the front page and the calendar itself is embedded near the bottom of the front page. Double-check the calendar events for details on which zoom room or physical location the office/help room hours are being held in. The information is also pinned in the cmse201-S24-help channel on Teams!

---
<a id="part_1"></a>

## Part 1. Doing basic math with Python (6 points)

[Back to Top](#toc)

As you explored with your order of magnitude modeling assignment in the Day 2 in-class assignment, Python is great at doing basic math -- much like your favorite calculator! As you get more comfortable with writing Python code, you might find yourself writing a bit of code when you need to do some calculations instead of pulling digging out your calculator or launching the calculator app on your phone. 

Doing basic math with Python involves using various "**operators**" -- these operators are the Python symbols used to do the actual math. For example, to add two variables, we use the "`+`" operator. To subtract, we use the "`-`" operator.

There are a number of other arithmetic (math) operators as well.

### &#9989;&nbsp; Question 1.1 (4 points)

In the markdown cell below, fill out the following table to indicate what math each operator is used for. If you don't know the answer off the top of your head, you should practice using Google (or your search engine of choice) to pick up some new Python knowledge (something even the pros do!). **Make sure you use the keyword "python" in your search terms when looking for these math operators.**

**Remember**: You can double-click on a markdown cell to edit it.

| Arithmetic Operator Symbol |  Name of Mathematical Operation | Description |
| :------------------------- | :------------------------------ | :---------- | 
| +                          | Addition                        | Add two variables or numbers together |
| -                          | Subtraction                     | Subtract one variable or number from another |
| *                          | Multiplication                  |  |
| /                          | ratio/division                  |  |
| **                         | Exponents/power                 |  |
| %                          | Remainder                       |  |

### &#9989;&nbsp; Question 1.2 (2 Points)

One arithmetic operator is missing from the table above. This is the "Floor Division" operator, which looks like this: `//`

In your own words in the cell below, explain what this operator does. In the code cell below that, use the operator in an example to confirm that your explanation is accurate.

The floor division operator disregards the numbers after the answers decimal point. So it only includes the "main" number.

In [3]:
# Put your example code that showcases how the // operator works to ensure that it agrees with your explanation

x = 3 // 2
print(x)

x = 3 / 2
print(x)

1
1.5


---
<a id="part_2"></a>

## Part 2: Finding common divisors (4 points)

[Back to Top](#toc)

For this question you should write code using the tools you learned so far. The code will find all the integers in the interval from 1 to 12000 that are **divisible by both 3 and 14**. **Store** all the numbers that satisfy this criterion into a _list_ and **print** the entire list. 

**Hint**: Remember the logical operators `and`, `or` and `not` are available in Python.

In [4]:
num_list = []

for i in range(12000):
    if i % 3 == 0:
        num_list.append(i)
    elif i % 14 == 0:
        num_list.append(i)

print(num_list)

[0, 3, 6, 9, 12, 14, 15, 18, 21, 24, 27, 28, 30, 33, 36, 39, 42, 45, 48, 51, 54, 56, 57, 60, 63, 66, 69, 70, 72, 75, 78, 81, 84, 87, 90, 93, 96, 98, 99, 102, 105, 108, 111, 112, 114, 117, 120, 123, 126, 129, 132, 135, 138, 140, 141, 144, 147, 150, 153, 154, 156, 159, 162, 165, 168, 171, 174, 177, 180, 182, 183, 186, 189, 192, 195, 196, 198, 201, 204, 207, 210, 213, 216, 219, 222, 224, 225, 228, 231, 234, 237, 238, 240, 243, 246, 249, 252, 255, 258, 261, 264, 266, 267, 270, 273, 276, 279, 280, 282, 285, 288, 291, 294, 297, 300, 303, 306, 308, 309, 312, 315, 318, 321, 322, 324, 327, 330, 333, 336, 339, 342, 345, 348, 350, 351, 354, 357, 360, 363, 364, 366, 369, 372, 375, 378, 381, 384, 387, 390, 392, 393, 396, 399, 402, 405, 406, 408, 411, 414, 417, 420, 423, 426, 429, 432, 434, 435, 438, 441, 444, 447, 448, 450, 453, 456, 459, 462, 465, 468, 471, 474, 476, 477, 480, 483, 486, 489, 490, 492, 495, 498, 501, 504, 507, 510, 513, 516, 518, 519, 522, 525, 528, 531, 532, 534, 537, 540, 543, 54

---
<a id="part_3"></a>

## Part 3: Fibonacci numbers (14 points)

[Back to Top](#toc)

Why is it that the number of petals in a flower is often one of the following numbers: 3, 5, 8, 13, 21, 34 or 55? For example, the lily has three petals, buttercups have five, the chicory has 21, the daisy has often 34 or 55 petals, etc. https://thefibonaccisequence.weebly.com/flowers.html

Are these numbers the product of chance? No! They all belong to the Fibonacci sequence: 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, ..., where each number is obtained from the sum of the two preceding.  The $n$th Fibonacci number $f_n$ can be derived by $f_n = f_{n-1} + f_{n-2}$. For a long time, it had been noticed that these numbers were important in nature, but only relatively recently did we understand that this is related to efficiency during the growth process of plants. 

### &#9989;&nbsp; Question 3.1 (5 Points) 
In this exercise, you are given a numerical sequence that has the form $1, 1, 2, 3, 5, \ldots$. Your goal is to write a __*loop*__ that computes 50 terms of the Fibonacci sequence. Store this sequence as a __*list*__ and then **print the entire list all at once.**. 

```
[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817, 39088169, 63245986, 102334155, 165580141, 267914296, 433494437, 701408733, 1134903170, 1836311903, 2971215073, 4807526976, 7778742049, 12586269025, 20365011074]
```

**Note:** If you Google Fibonacci number python code you will likely find something about _recursion_. **You are not allowed to use recursion!**. You should be able to write this code using only loops and lists! No functions needed.


In [5]:
fib_list = [1,1]

for i in range(50):
    if i > 0:  
        fib_num = fib_list[i] + fib_list[i-1]
        fib_list.append(fib_num)
        
print(fib_list)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817, 39088169, 63245986, 102334155, 165580141, 267914296, 433494437, 701408733, 1134903170, 1836311903, 2971215073, 4807526976, 7778742049, 12586269025, 20365011074]


### &#9989;&nbsp; Question 3.2 (5 Points)  
Once you have your list of Fibonacci numbers, use it to calculate the ratios $R_n = f_{n}/f_{n+1}$ and store it into a new list. 
At the end print the entire new list all at once. 

In [6]:
new_fib = []

for i in range(len(fib_list)):
    if i != (len(fib_list) - 1):
        result = fib_list[i] / fib_list[i+1]
        new_fib.append(result)

print(new_fib)

[1.0, 0.5, 0.6666666666666666, 0.6, 0.625, 0.6153846153846154, 0.6190476190476191, 0.6176470588235294, 0.6181818181818182, 0.6179775280898876, 0.6180555555555556, 0.6180257510729614, 0.6180371352785146, 0.6180327868852459, 0.6180344478216818, 0.6180338134001252, 0.6180340557275542, 0.6180339631667066, 0.6180339985218034, 0.618033985017358, 0.6180339901755971, 0.6180339882053251, 0.618033988957902, 0.6180339886704432, 0.6180339887802427, 0.618033988738303, 0.6180339887543226, 0.6180339887482036, 0.6180339887505408, 0.6180339887496481, 0.618033988749989, 0.6180339887498588, 0.6180339887499086, 0.6180339887498896, 0.6180339887498969, 0.6180339887498941, 0.6180339887498951, 0.6180339887498948, 0.6180339887498949, 0.6180339887498948, 0.6180339887498949, 0.6180339887498948, 0.6180339887498949, 0.6180339887498949, 0.6180339887498949, 0.6180339887498949, 0.6180339887498949, 0.6180339887498949, 0.6180339887498949, 0.6180339887498949]


### &#9989;&nbsp; Questions 3.3 (6 Points) 

1. (2 points) Do you notice something interesting in the printed numbers? How do the numbers change? Try googling the last number.
2. (4 points) What type of loops did you use to write the above programs, **looping by value** or **looping by index**? If you used looping by value, can you re-write your code using looping by index? Instead, if you used looping by index, can you re-write your code using looping by value?

You must explain your answers. 

**Note:** If you use code from the internet of AI tools you must cite the source by including a comment in your code (using the `#` symbol) that provides the URL that your code came from and the date you accessed that URL. If you used AI tools don't forget to add the prompt used. You must explain any new code (_e.g._ built-in python functions) that has not been covered in class. 


In [7]:
fibo_list = []
a, b = 0, 1

while a <= 50:
    fibo_list.append(a)
    a, b = b, a + b
    
print(fib_list)
#Created with the aid of chatGPT.
#Prompt: can you create a python loop that generates fibonacci numbers and puts them in a list by value instead of by index

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817, 39088169, 63245986, 102334155, 165580141, 267914296, 433494437, 701408733, 1134903170, 1836311903, 2971215073, 4807526976, 7778742049, 12586269025, 20365011074]


1. The last number in my new_fib list asymptotically approached the number called the "golden ratio." It is a prevalent number in aesthetically pleasing mathematical properties, appearing in art, architecture, and other fields.

2. I chose to loop by index because it was the most similar to the formula's provided. I will try to loop by value.

---
<a id="part_4"></a>

## Part 4. Amortizing Loans (16 points)

[Back to Top](#toc)


An amortizing loan is a type of loan in which the borrower makes regular payments consisting of both principal and interest. These payments are designed to fully repay the loan by the end of its term. One common example of an amortizing loan is a fixed-rate mortgage for buying a home.

An amortization schedule is a detailed table that outlines each payment on an amortizing loan, including the portion that goes toward paying off the principal amount and the portion that covers the interest. This schedule helps borrowers understand how their loan balance decreases over time. You can find more details [here](https://www.investopedia.com/terms/a/amortization.asp#toc-what-is-amortization) if you are interested. 

### &#9989;&nbsp; Question 4.1 (12 points)

Your task is to create a Python program that generates an amortization schedule for a loan of \$30,000 with an annual interest rate of 3% over a 4-year period. The program should print out the amortization schedule, including the amount of interest, the amount of payment towards the principal, and the remaining loan balance.

**Formula for Monthly Payment:**

The monthly payment on an amortizing loan can be calculated using the following formula:

$$ M = \frac{P \cdot r \cdot (1 + r)^N}{(1 + r)^N - 1} $$

where
- $M$ is the monthly payment.
- $P$ is the principal loan amount (in this case, \$ 30,000).
- $r$ is the monthly interest rate, the annual interest rate divided by the number of months in a year.
- $N$ is the total number of monthly payments (in this case, the loan term in months).

The amount of interest paid in each monthly installment is obtained by multiplying the remaining balance times the monthly interest rate. 

**Sample Output:**

Your program should generate an amortization schedule with numbers rounded to 2 decimal places. Here is a sample output

| Month  | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |
|--------|--------------------|-------------------------------------|-------------------|
|        |                    |                                     | 30,000            |
| 1      | 75                 | 589.03                              | 29,410.97         |
| 2      | 73.53              | 590.50                              | 28,820.47         |
| 3      | 72.05              | 591.98                              | 28,228.49         |
| 4      | 70.57              | 593.46                              | 27,635.03         |

Some started code is given below.

**IMPORTANT:** If you use a `for` loop in Part 2, then you must use a `while` loop in here and viceversa. Otherwise you will receive only half the points.


In [8]:
P = 30000
r = .03 / 12
N = 12
month = 0

M = round((P * r * (1 + r)**N) / ((1 + r)**N - 1),2)

while month < 4:
    
    month += 1
    P = round(P - M,2)
    month_interest = round(r * P, 2)
    
    print("| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |")
    print()
    print("|  ", month, "   |","|  ", M, "   |", "|  ", month_interest, "   |", "|  ", round(M + month_interest,2), "   |", "|  ", P, "   |")
    print()


| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   1    | |   2540.81    | |   68.65    | |   2609.46    | |   27459.19    |

| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   2    | |   2540.81    | |   62.3    | |   2603.11    | |   24918.38    |

| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   3    | |   2540.81    | |   55.94    | |   2596.75    | |   22377.57    |

| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   4    | |   2540.81    | |   49.59    | |   2590.4    | |   19836.76    |



### &#9989;&nbsp; Questions 4.2 (4 Points) 

What type of loops did you use to write the above programs, **looping by value** or **looping by index**? If you used looping by value, can you re-write your code using looping by index? Instead, if you used looping by index, can you re-write your code using looping by value?

You must explain your answers. 

**Note:** If you use code from the internet of AI tools you must cite the source by including a comment in your code (using the `#` symbol) that provides the URL that your code came from and the date you accessed that URL. If you used AI tools don't forget to add the prompt used. You must explain any new code (_e.g._ built-in python functions) that has not been covered in class. 


The original problem looped by value (using while loops) through all the months of the amatorization period then performed all the necessary calculations, month by month. In particular, ensuring that the remaining balance gradually decreases along with the monthly interest, contributing to the payment towards pricipal.

In [9]:
P = 30000
r = .03 / 12
N = 12

M = round((P * r * (1 + r)**N) / ((1 + r)**N - 1),2)

for i in range(4): #Now loops through the months by index instead of value.
    
    P = round(P - M,2)
    month_interest = round(r * P, 2)
    
    print("| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |")
    print()
    print("|  ", i + 1, "   |","|  ", M, "   |", "|  ", month_interest, "   |", "|  ", round(M + month_interest,2), "   |", "|  ", P, "   |")
    print()


| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   1    | |   2540.81    | |   68.65    | |   2609.46    | |   27459.19    |

| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   2    | |   2540.81    | |   62.3    | |   2603.11    | |   24918.38    |

| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   3    | |   2540.81    | |   55.94    | |   2596.75    | |   22377.57    |

| Month | Monthly Payment | Amount of Interest | Amount of Payment Towards Principal | Remaining Balance |

|   4    | |   2540.81    | |   49.59    | |   2590.4    | |   19836.76    |



---
<a id="part_5"></a>

## Part 5. Dictionaries and everything else (7 points)

[Back to Top](#toc)

Occasionally during this class you will be expected to learn a new Python concept "on-the-fly". This is when it becomes important for you to be able to digest Python documentation and do some intelligent Googling. For this part of the assignment, you're going to see if you can pick up a bit about how another Python data container works, the **dictionary**.

Dictionaries are like lists in the sense that we have some object (e.g. integer, string, etc.) occupying a position in the dictionary. However, dictionaries differ in that we refer to each element by a **key** rather than an integer index. For example, in the dictionary `example_dict` in the cell below, the key string `"red"` corresponds to the **value** integer `-2`, the key string `"green"` corresponds to the value string `"beach_ball"`. We can include values of different types in the same dictionary, just as we did with lists above.

__*IMPORTANT:*__ Keys may be of any type, but an element can only be accessed by its specific key. Even though `"red"` is the first key, `example_dict[0]` is not a valid way to access it. Also note that dictionaries are written using **curly braces `{}`**; lists use **square brackets `[]`**. The key-value pairs are separated by a **colon `:`**.

In [10]:
# Here is how to define a dictionary in Python with strings as keys
example_dict = {"red": -2, "orange": "Cata", "green": "beach_ball", "blue": 25}

# This outputs the value that corresponds to the key "red"
example_dict["red"]

# The following line will error, because integer 1 is not a valid key for this dictionary
example_dict[1]


KeyError: 1

In addition to the above example, here is a table that highlights some of the other properties of dictionaries in Python (you may need to scroll side-to-side):

|Container Type $\hspace{0.5in}$|Mutable or Immutable $\hspace{0.5in}$|Initialization *Without* Values $\hspace{0.9in}$|Initializtion *With* Values $\hspace{2.5in}$|Adding Values to Container $\hspace{3.5in}$|Removing Values from Container $\hspace{2.0in}$|Modifying Values $\hspace{2.0in}$|Access Method $\hspace{1.0in}$|Notable Operations and Additional Information $\hspace{2.5in}$|
|:---|:---|:---|:---|:---|:---|:---|:---|:---|
|**Dictionary**|Mutable| `student={}` | `>>> student={'name': 'John Doe', 'age': 22, 'college': 'MSU'}` | `>>> student['major']='Computer Science'` <br> `>>> student` <br> `{'name': 'John Doe', 'age': 22, 'college': 'MSU', 'major': 'Computer Science'}` | `del dictName[keyName] #This method removes all entries associated with the given key` | `>>> student['age'] = 23` <br> `>>> student` <br> `{'name': 'John Doe', 'age': 23, 'college': 'MSU', 'major': 'Computer Science'}`| Access by key word. Note that this key **must** be a string. <br> `>>>student['college']` <br> `MSU`| The 'in' keyword can be very helpful with dictionaries. Ex: <br><ul><li>`'k' in dict #Returns true if key 'k' is in dictionary dict`</li><li>`'k' not in dict #Returns true if key 'k' is not in dicitonary dict`</li><li>`for key in dict #This will iterate over all keys in dictionary dict`</li></ul> <br>See webpage [here](http://www.python-course.eu/python3_dictionaries.php) for additional helpful methods and operations|

### &#9989;&nbsp; Question 5.1 (3 Points) 

**Define a dictionary called `cmse_201`**. There will be 4 elements in your dictionary. The keys should be the strings `"Instructor"`, `"TA"`, `"Section"`, and `"Class Days"`. The values assigned to each key should be your respective instructor, TA, section number (e.g. 1, 2, 3, or 5), and a _list_ that contains the days of the week that the course meets. Once you've created it, print your dictionary.


In [ ]:

cmse_201 = {"Instructor" : "Bao, Lianzhang", "TA" : "Maryam Berijanian", "Section" : "001", "Class Days" : ["Monday", "Wednesday"]}
print(cmse_201)

### &#9989;&nbsp; Question 5.2 (4 Points) 

As you can see dictionaries can be a great way to store data. Assume you live in a time where streaming services did not exist and you actually had hard copies of movies (DVD, VideoCassettes, etc..), music albums (Music cassette, CD, Vinyl, etc...), and books (ever heard of a library?). Now imagine you work for Blockbuster, or a record store, or a library and you have to make a catalog of their collection. 

**Write** a dictionary whose keys are (choose one):

1. **Records Store Music:** by storing artist' name(s), albums, release year, genre(s), type of technology (Vinyl, CD, music casette), aisle number, shelf number
2. **Blockbuster Movies:** by storing movie title, release year, director, cast, technology (DVD, Bluray, VC, IMAX, etc.), IMDB rating, genre(s), aisle number, shelf number
3. **Library Books:** by storing book author, title, publisher, year, genre(s), type of cover (hardcover, paperback), aisle number, shelf number

Now, store 5 to 10 of your favorite (directors, artists, authors) and for each store 3 or more titles (movies, albums, books). If you don't have favorites you can pick them from the internet.

**Note** that each dictionary has the keys **aisle number** and **shelf number**. These indicate where the item is stored so that if a customer asks, you (as the clerk) can easily find the item for them. You can choose these however you want, _e.g._ Aisle 1, Shelf 2

In [ ]:
Blockbuster_Movies = {"Movie Title" : "Stair Wars", "Release Year": 1822, "Director" : "Mr.DirectorMan", "Cast" : "Spongebob & Patrick", "Technology" : "VCR", "IMDB Rating" : "3/10", "Genre" : "Horror", "Aisle Number" : 7000, "Shelf Number" : 800}

In [ ]:
print (Blockbuster_Movies["Movie Title"])

---
<a id="part_6"></a>

## Part 6. Conditional statements and debugging (6 points): 
[Back to Top](#toc)

In the code cell below there is an complete working code that does not throw an error, but it does have a bug. The code is supposed to assign a letter grade to each student based on this rubric

        A if greater or equal than 90
        B if greater or equal than 80 but less than 90
        C if greater or equal than 70 but less than 80
        D if greater or equal than 60 but less than 70
        F if less than 60

At the end the code stores the letter grade into the list `students_letter_grades`.
However, it gives the wrong answer. Giulia should get a "C" not a "D"


In [ ]:
students_names = ['Giuseppe', 'Giulia', 'Marco',
                  'Isabella', 'Mario', 'Federica',
                  'Alessandro', 'Valentina', 'Davide',
                  'Eleonora', 'Antonio','Martina', 
                  'Roberto', 'Francesca', 'Lorenzo']

students_grades = [50, 72, 74, 
                   67, 100, 96,
                   97, 86, 70,
                   74, 76, 93,
                   96, 87, 82]

students_letter_grades = [] 

for index in range(len(students_grades)):
    #print(index)
    
    grade = students_grades[index] 
    
    name = students_names[index]
    
    if grade >= 90:
        letter_grade = "A"
    
    if grade >= 80:
        letter_grade = "B"
    
    if grade >= 70:
        letter_grade = "C"
    
    if grade >= 60:
        letter_grade = "D"
    
    if grade < 60:
        letter_grade = "F"
    
    students_letter_grades.append( name + "_" + letter_grade )

print(students_letter_grades)

In [ ]:
    #Number 3 from below
    
    for index in range(len(students_grades)):
    #print(index)
    
        grade = students_grades[index] 

        name = students_names[index]
        
        if grade >= 90:
            letter_grade = "A"

        # Switching the following code to elif(s) instead of if(s) fixed the bug

        elif grade >= 80:
            letter_grade = "B"

        elif grade >= 70:
            letter_grade = "C"

        elif grade >= 60:
            letter_grade = "D"

        elif grade < 60:
            letter_grade = "F"

        students_letter_grades.append( name + "_" + letter_grade )

print(students_letter_grades)

### &#9989;&nbsp; **Question 6.1 (6 points):** 

In the cell below, answer the following questions

1. A plain english description of the error in the code and why you get the error. (2 points)
2. How would you fix this bug ? (2 points) 
3. Create a new cell and write your solution in there. (2 points)

1. In the conditional statements, each iteration was run through all the if statements. If the statement was true, it moves on to the next Giulia happened to also meet the criteria grade >= 60 after meeting grade >= 70.

2. Simply changing the if statements following the first one to elif makes this code a choice rather than a runthrough. Once something is true, no further statements in this chain are executed.

---
## Wait! Before you submit your notebook, read the following important recommendations

When your TA opens your notebook to grade the assignment, it will be really useful if your notebook is saved in a fully executed state so that they can see the output of all your code. Additionally, it may be necessary from time to time for the TA to actually run your notebook, but when they run the notebook, it is important that you are certain that all the code will actually run for them!

You should get into the following habit: **before you save and submit your final notebook for your assignments, go to the "Kernel" tab at the top of the notebook and select "Restart and Run all".** This will restart your notebook and try to run the code cell-by-cell from the top to the bottom. Once it finished, review you notebook and make sure there weren't any errors that popped up. Sometimes, when working with notebooks, we accidentally change code in one cell that break our code in another cell. If your TA stumbles across code cells that don't work, **they will likely have to give you a zero for those portions of the notebook that don't work**. Testing your notebook one last time is a good way to make sure this doesn't happen.

**Once you've tested your whole notebook again, make sure you save it one last time before you upload it to D2L.**

---

### Congratulations, you're done! ###

Submit this assignment by uploading it to the course Desire2Learn web page. Go to the "Homework Assignments" section, find the submission folder link for Homework #1, and upload it there.

Copyright &#169; 2023, [Department of Computational Mathematics, Science and Engineering](https://cmse.msu.edu/) at Michigan State University, All rights reserved.